graph/graph.go

```diff
@@ -27,11 +28,67 @@ import (
 	"github.com/docker/docker/runconfig"
 )
 
+// The type is used to protect pulling or building related image
+// layers from deleteing when filtered by dangling=true
+// The key of layers is the images ID which is pulling or building
+// The value of layers is a slice which hold layer IDs referenced to
+// pulling or building images
+type retainedLayers struct {
+	layers map[string][]string
+	sync.Mutex
+}
+
+func (r *retainedLayers) Add(layerIDs []string, imgID string) {
+	r.Lock()
+	defer r.Unlock()
+	for _, layerID := range layerIDs {
+		if imgs, exists := r.layers[layerID]; exists {
```

Comment:
I think you can use just `r.layers[layerID] = append(r.layers[layerID], imgID)` instead of if/else.